nebullvm/operations/optimizations/base.py

```diff
@@ -0,0 +1,242 @@
+    ModelCompiler,
+    QuantizationType,
+    OptimizationTime,
+    ModelParams,
+    DeepLearningFramework,
+    ModelCompressor,
+)
+from nebullvm.tools.data import DataManager
+from nebullvm.tools.feedback_collector import FEEDBACK_COLLECTOR
+from nebullvm.tools.transformations import MultiStageTransformation
+
+
+class Optimizer(Operation, abc.ABC):
+    def __init__(self):
+        super().__init__()
+        self.optimized_models = []
+        self.source_dl_framework = None
+        self.pipeline_dl_framework = None
+        self.compiler_ops = {}
+        self.build_inference_learner_ops = {}
+        self.validity_check_op = MetricDropMeasure()
+
+    def execute(
+        self,
+        model: Any,
+        input_data: DataManager,
+        optimization_time: OptimizationTime,
+        metric_drop_ths: str,
+        metric: Callable,
+        model_params: ModelParams,
+        model_outputs: List[Tuple[Any, ...]],
+        ignore_compilers: List[ModelCompiler],
+        ignore_compressors: List[ModelCompressor],
+        source_dl_framework: DeepLearningFramework,
+    ):
+        self.source_dl_framework = source_dl_framework
+
+        # TODO: implement and select compressors from hardware
+
+        compilers = self._select_compilers_from_hardware()
+        (
+            self.compiler_ops,
+            self.build_inference_learner_ops,
+        ) = self._load_compilers(
+            ignore_compilers=ignore_compilers,
+            compilers=compilers,
+        )
+        self.optimize(
+            model=model,
+            input_data=input_data,
+            optimization_time=optimization_time,
+            metric_drop_ths=metric_drop_ths,
+            metric=metric,
+            model_params=model_params,
+            model_outputs=model_outputs,
+            ignore_compilers=ignore_compilers,
+        )
+
+    @abc.abstractmethod
+    def _select_compilers_from_hardware(self):
+        raise NotImplementedError()
+
+    def _load_compilers(
+        self,
+        ignore_compilers: List[ModelCompiler],
+        compilers: List[ModelCompiler],
+    ):
+        from nebullvm.operations.optimizations.optimizers import (
+            COMPILER_TO_OPTIMIZER_MAP,
+            COMPILER_TO_INFERENCE_LEARNER_MAP,
+        )
+
+        compiler_ops = {
+            compiler: COMPILER_TO_OPTIMIZER_MAP[compiler][
+                self.pipeline_dl_framework
+            ]()
+            for compiler in compilers
+            if compiler not in ignore_compilers
+            and compiler in COMPILER_TO_OPTIMIZER_MAP
+        }
+        build_inference_learner_ops = {
+            compiler: COMPILER_TO_INFERENCE_LEARNER_MAP[compiler][
+                self.pipeline_dl_framework
+            ]()
+            for compiler in compilers
+            if compiler not in ignore_compilers
+            and compiler in COMPILER_TO_OPTIMIZER_MAP
+        }
+
+        return compiler_ops, build_inference_learner_ops
+
+    def optimize(
```

Comment:
what's the goal of exposing this method as public? In what cases the user should use it?